### PR TITLE
Add dev fmt subcommand

### PR DESCRIPTION
**Pull request by mikerite**
changelog: none

### CODE REVIEW

**Comment by mikerite** [comment]
I forgot to say that this addresses #3835 
This commit doesn't include any component/toolchain checking but it should still be useful.

File: tests/fmt.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#[test]
+fn fmt() {
```
Comment (yaahallo) [inline tests/fmt.rs]: Does this mean that `./util/dev fmt --check` will be run as part of `cargo test`?

If so, does it output the correct command to run to fix formatting issues upon test failure?

Comment (mikerite) [reply tests/fmt.rs]: > Does this mean that ./util/dev fmt --check will be run as part of cargo test?

Yes

> If so, does it output the correct command to run to fix formatting issues upon test failure?

It tells you to run `./util/dev fmt` which should fix it.

Comment (mikerite) [reply tests/fmt.rs]: Actually, it's part of the stdout that gets written out when the test fails. I guess I could add it in the assert.

**Comment by flip1995** [comment]
LGTM! With the `tests/fmt.rs` test, we could probably clean up the `ci/base-tests.sh`.

Are you still motivated to add the toolchain+component checking in this PR or should we merge this now and include that later?

File: tests/fmt.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    if option_env!("RUSTC_TEST_SUITE").is_some() {
+        return;
+    }
+
+    let root_dir = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    let dev_dir = root_dir.join("clippy_dev");
+    let output = std::process::Command::new("cargo")
+        .current_dir(dev_dir)
+        .args(&["run", "--", "fmt", "--check"])
```
Comment (flip1995) [inline tests/fmt.rs]: Should we specify here to run nightly `rustfmt`?

```suggestion
        .args(&["+nightly", "run", "--", "fmt", "--check"])
```

Comment (flip1995) [reply tests/fmt.rs]: Never mind, this is already specified in `cargo_fmt(..)` and `rustfmt(..)`.

**Comment by mikerite** [comment]
> Are you still motivated to add the toolchain+component checking in this PR or should we merge this now and include that later?

Let's merge this now.

**Comment by flip1995** [comment]
Everything from here can also be removed:
https://github.com/rust-lang/rust-clippy/pull/4232/files#diff-ce18faa8208e5babb77318549d5c101fL54

**Comment by flip1995** [comment]
Great work, thanks!

@bors r+

**Comment by flip1995** [comment]
https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/25748824#L854

Oh no. Didn't you work on this same problem for dogfood @mikerite?

**Comment by mikerite** [comment]
Yes, it's the same as the problem we had with dogfood. Specifying the version of cargo to use should fix it.

**Comment by phansch** [comment]
Looks like we'll have to replace `ONCE_INIT` in the UI tests?

**Comment by flip1995** [comment]
Removed the merge commits

@bors r+ p=10

**Comment by lzutao** [comment]
Need to add `rustup component add rustfmt --toolchain nightly` in appveyor.

**Comment by mikerite** [comment]
This is a rustfmt bug.

```
PS C:\Projects\rust-clippy> rustfmt .\tests\ui\implicit_hasher.rs
error: couldn't read \\?\C:\Projects\rust-clippy\tests\ui\..\auxiliary\test_macro.rs: The filename, directory name, or volume label syntax is incorrect. (os error 123)
  --> \\?\C:\Projects\rust-clippy\tests\ui\implicit_hasher.rs:88:9
   |
88 | pub mod test_macro;
   |         ^^^^^^^^^^
```
It's either rust-lang/rustfmt#1873 or related to it. I'm going to add windows specific excludes for files like that.

**Pull request merged**